pkg/rc/replication_controller.go

```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
 	}
 
 	return nil
 }
 
-func (rc *replicationController) transferPetNodes(ineligibleCurrent []types.NodeName) error {
-	errMsg := fmt.Sprintf("RC has scheduled %d ineligible nodes: %s", len(ineligibleCurrent), ineligibleCurrent)
-	err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+func (rc *replicationController) watchHealth() {
+	defer func() {
+		rc.isWatchingHealth = false
+		rc.quitTransfer()
+	}()
+
+	rc.mu.Lock()
+	podID := rc.Manifest.ID()
+	rc.mu.Unlock()
+
+	resultCh := make(chan health.Result)
+	errCh := make(chan error)
+	healthQuitCh := make(chan struct{})
+
+	isHealthy := false
+	rc.healthChecker.WatchNodeService(rc.nodeTransfer.newNode, podID.String(), resultCh, errCh, healthQuitCh)
+
+	for {
+		select {
+		case err := <-errCh:
+			rc.logger.WithError(err).Errorln("Node transfer health checker sent error")
+		case currentHealth := <-resultCh:
+			if currentHealth.Status == health.Passing {
+				err := rc.finishTransfer()
+				if err != nil {
+					rc.logger.WithError(err).Errorln("Could not complete transfer transaction")
+				}
+				return
+			}
+		case <-rc.nodeTransfer.quit:
+			close(healthQuitCh)
+			close(errCh)
+			if isHealthy {
+				err := rc.finishTransfer()
+				if err != nil {
+					rc.logger.WithError(err).Errorln("Could not complete transfer transaction")
+				}
+			}
+			return
+		}
+	}
+}
+
+func (rc *replicationController) finishTransfer() error {
+
+	current, err := rc.CurrentPods()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	txn, cancelFunc := rc.newAuditingTransaction(ctx, current.Nodes())
+	defer cancelFunc()
+
+	err = rc.nodeTransfer.unlocker.UnlockTxn(ctx)
+	if err != nil {
+		return err
+	}
+
+	txn.AddNode(rc.nodeTransfer.newNode)
+
+	rc.mu.Lock()
+	man := rc.Manifest
+	rc.mu.Unlock()
+
+	labelKey := labels.MakePodLabelKey(rc.nodeTransfer.newNode, man.ID())
+	err = rc.podApplicator.SetLabelsTxn(ctx, labels.POD, labelKey, rc.computePodLabels())
+	if err != nil {
+		return err
+	}
+
+	err = rc.unschedule(txn, rc.nodeTransfer.oldNode)
+	if err != nil {
+		return err
+	}
+
+	ok, resp, err := txn.CommitWithRetries(rc.txner)
+	switch {
+	case err != nil:
+		errMsg := fmt.Sprintf(
+			"transfer unlock, label, and unschedule txn returned err after timeout: %s",
+			err,
+		)
+		alertErr := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+		if alertErr != nil {
+			rc.logger.WithError(err).Errorln("Unable to send alert")
+		}
+		return err
+	case !ok:
+		errMsg := fmt.Sprintf(
+			"Transaction violation trying to unlock node transfer session and label new node. New RC may have scheduled node: %s",
+			transaction.TxnErrorsToString(resp.Errors),
+		)
+		err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+		if err != nil {
+			rc.logger.WithError(err).Errorln("Unable to send alert")
+		}
```

Comment:
do we need to delete the node transfer here? we probably do otherwise the new node will be ignored potentially forever right?

Reply:
It gets deleted by `quitTransfer` which is called by a defer in `watchHealth`, which is the function that calls this `finishTransfer` here. Your question is just further evidence in support of your previous comment. Something should be done about the organization of these functions to make this more clear and explicit